src/trajectory.jl

```diff
@@ -218,190 +282,167 @@ Base.show(io::IO, d::Termination) = print(io, "Termination(dynamic=$(d.dynamic),
 Base.:*(d1::Termination, d2::Termination) = Termination(d1.dynamic || d2.dynamic, d1.numerical || d2.numerical)
 isterminated(d::Termination) = d.dynamic || d.numerical
 
-"""
-A full binary tree trajectory with only necessary leaves and information stored.
-"""
-struct FullBinaryTree{S<:AbstractTreeSampler}
-    zleft       # left most leaf node
-    zright      # right most leaf node
-    zcand       # candidate leaf node
-    sampler::S  # condidate sampler
-    termination # termination reasons
-    α           # MH stats, i.e. sum of MH accept prob for all leapfrog steps
-    nα          # total # of leap frog steps, i.e. phase points in a trajectory
-end
-
-"""
-Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
-"""
-function isturn(h::Hamiltonian, zleft::PhasePoint, zright::PhasePoint)
-    θdiff = zright.θ - zleft.θ
-    s = (dot(θdiff, ∂H∂r(h, zleft.r)) >= 0 ? 1 : 0) * (dot(θdiff, ∂H∂r(h, zright.r)) >= 0 ? 1 : 0)
-    return Termination(s == 0, false)
-end
-# function isturn(h::Hamiltonian, zleft::PhasePoint, zright::PhasePoint)
-#     θdiff = zright.θ - zleft.θ
-#     s = (dot(-θdiff, ∂H∂r(h, zleft.r)) < 0) && (dot(θdiff, ∂H∂r(h, zright.r)) < 0)
-#     return Termination(s, false)
-# end
-
 """
 Check termination of a Hamiltonian trajectory.
 """
-isterminated(
+Termination(
     s::SliceTreeSampler,
     nt::NUTS,
     H0::F,
     H′::F
 ) where {F<:AbstractFloat} = Termination(false, !(s.logu < nt.Δ_max + -H′))
-isterminated(
+
+"""
+Check termination of a Hamiltonian trajectory.
+"""
+Termination(
     s::MultinomialTreeSampler,
     nt::NUTS,
     H0::F,
     H′::F
 ) where {F<:AbstractFloat} = Termination(false, !(-H0 < nt.Δ_max + -H′))
 
 """
-    combine(h::Hamiltonian, tleft::FullBinaryTree, tright::FullBinaryTree)
+A full binary tree trajectory with only necessary leaves and information stored.
+"""
+struct FullBinaryTree{C<:AbstractTerminationCriterion}
+    zleft       # left most leaf node
+    zright      # right most leaf node
+    c::C        # termination criterion
+    α           # MH stats, i.e. sum of MH accept prob for all leapfrog steps
```

Comment:
Perhaps `α` ==> `sum_α` for clarity?

Reply:
How about `α_acc` (accumulated α)? Also we should change `n` to `n_steps` or something?

Reply:
I feel `sum_α` is more readable. Maybe change `n` to `n_α`?

Reply:
You mean `n` to `sum_n`?

Reply:
No, `sum_n` is more confusing than `n` I think. Maybe we can use `n` for now?

Reply:
Sure.